File: go/2024/solutions/day01.go
package solutions

import (
	"fmt"
	"math"
	"slices"
	"strconv"
	"strings"
)

type Day01 struct{}
type input01 struct {
	Left  []int
	Right []int
}

func (d Day01) Parse(raw string) input01 {
	lines := strings.Split(raw, "\n")
	left := make([]int, len(lines))
	right := make([]int, len(lines))
	for _, line := range lines {
		nums := strings.Split(line, "   ")

		l, _ := strconv.Atoi(nums[0])
		left = append(left, l)

		r, _ := strconv.Atoi(nums[len(nums)-1])
		right = append(right, r)
	}
	return input01{left, right}
}

func (d Day01) Part1(raw string) {
	fmt.Println("Day 1 Part 1")
	input := d.Parse(raw)
	slices.Sort(input.Left)
	slices.Sort(input.Right)

	dist := 0
	for i := 0; i < len(input.Left); i++ {
		dist += int(math.Abs(float64(input.Left[i] - input.Right[i])))
	}
	fmt.Printf("Solution: %d\n", dist)
}

func (d Day01) Part2(raw string) {
	fmt.Println("Day 1 Part 2")
	input := d.Parse(raw)
	total := 0
	for _, valL := range input.Left {
		count := 0
		for _, valR := range input.Right {
			if valL == valR {
				count++
			}
		}
		total += valL * count
	}
	fmt.Printf("Solution: %d\n", total)
}

File: rust/2024/src/solutions/day_01.rs
use std::iter::zip;

use itertools::Itertools;
use pathfinding::num_traits::abs;

use super::day_trait::Day;

#[derive(Debug)]
pub struct Input {
    lines: Vec<String>,
    left: Vec<isize>,
    right: Vec<isize>,
}

fn parse_input(raw: &str) -> Input {
    let lines: Vec<String> = raw.split("\n").map(|s| s.to_owned()).collect();
    let mut left = Vec::new();
    let mut right = Vec::new();
    for line in &lines {
        let (a, b) = line.split_once("   ").unwrap();
        left.push(a.parse().unwrap());
        right.push(b.parse().unwrap());
    }
    Input { lines, left, right }
}

pub struct Day01;
impl Day for Day01 {
    fn day(&self) -> u8 {
        01
    }

    fn part_1(&self, raw: &str) {
        println!("Day {} part 1", self.day());
        let input: Input = parse_input(raw);

        let sum: isize = zip(input.left.iter().sorted(), input.right.iter().sorted())
            .map(|(a, b)| abs(a - b))
            .sum();

        println!("Answer is {}", sum);
    }

    fn part_2(&self, raw: &str) {
        println!("Day {} part 2", self.day());
        let input: Input = parse_input(raw);

        let sum: isize = input
            .left
            .iter()
            .map(|l| input.right.iter().filter(|&r| r == l).count() as isize * l)
            .sum();

        println!("Answer is {}", sum);
    }
}
